### PR TITLE
Tobler + OSM

### PR DESCRIPTION
**Pull request by fosskers**
### TODO

- [x] Spruce config
- [x] `decline` for CLI opts
- [x] Overlay OSM geometries before Reprojecting
- [x] Testing

Closes #2

### CODE REVIEW

File: src/main/scala/ingest/PyramidTobler.scala [hunk 1]
```diff
@@ -85,29 +109,89 @@ object ToblerPyramid {
...
+        case Right((ns, ws, _)) => {
+          val roads: RDD[(Long, osm.Way)] = ws.filter { case (_, w) => w.meta.tags.contains("highway") }
+
+          val lines: RDD[Line] = osm.toHistory(ns, roads)._2.map(_.geom)
+
+          /* Silently dumps the `Metadata` portion of the returned value.
+           * It's just a `LayoutDefinition` that we borrowed from `tobler`
+           * anyway, so we don't need it.
+           */
+          val geomTiles: RDD[(SpatialKey, Tile)] = lines.rasterize(1, BitCellType, tobler.metadata.layout)
```
Comment (fosskers) [inline src/main/scala/ingest/PyramidTobler.scala]: What `CellType` is best to use here?

Comment (fosskers) [inline src/main/scala/ingest/PyramidTobler.scala]: What `CellType` and value are best to use here?

File: src/main/scala/ingest/PyramidTobler.scala [hunk 2]
```diff
@@ -85,29 +109,89 @@ object ToblerPyramid {
             }
         }
 
+      /* ORC file is assumed to be a snapshot with the most recent versions of every Element. */
+      osm.fromORC(orc) match {
+        case Left(err) => println(err)
+        case Right((ns, ws, _)) => {
+          val roads: RDD[(Long, osm.Way)] = ws.filter { case (_, w) => w.meta.tags.contains("highway") }
+
+          val lines: RDD[Line] = osm.toHistory(ns, roads)._2.map(_.geom)
```
Comment (fosskers) [inline src/main/scala/ingest/PyramidTobler.scala]: For the moment, `toHistory` has the more convenient return type. Luckily it won't do extra work even if the ORC file given to it wasn't a history file.

File: terraform/emr.tf [hunk 3]
```diff
@@ -2,7 +2,7 @@ resource "aws_emr_cluster" "emr-spark-cluster" {
   name          = "geotrellis-friction-surface Ingest"
   applications  = ["Hadoop", "Spark", "Ganglia"]
   log_uri       = "${var.s3_uri}"
-  release_label = "emr-5.7.0"
+  release_label = "emr-5.10.0"
```
Comment (fosskers) [inline terraform/emr.tf:5]: As of November.

**Comment by fosskers** [comment]
My first ingest scoped to California succeed, generating 12gb of tiles, around ~80k of them in total. It took just over an hour to complete on EMR w/ a 50 node cluster.

**Comment by fosskers** [comment]
Attempting to inspect them visually with [Trelliscope](https://github.com/rajadain/trelliscope).

**Comment by fosskers** [comment]
After altering the code to do cell-wise tile merging, here's what we see:

2017-11-30-145914_1918x1078_scrot

That's not a very high zoom, but it's as deep as I seem to be able to go with the current GPS setup. Any deeper creates a spark job that takes too long, timing out the original tile request.

I'll try it next with water bodies included.

**Comment by fosskers** [comment]
Slightly better image at lower zoom.

2017-12-01-085952_1918x1078_scrot

**Comment by fosskers** [comment]
@jpolchlo helped me speed up the tile visualization. We can definitely see the reduced friction effect of roads at high zooms:

2017-12-01-101919_1918x1078_scrot

**Comment by jamesmcclain** [comment]
The yellow blocks are not understood, but if they turn out to be a problem they can be fixed later.  :+1:

**Review by jamesmcclain — APPROVED** [review]

**Pull request merged**